### ISSUE TITLE
Support HTTP calls in flutter_test

### ISSUE THREAD

**Issue by tudor07** [flutter/flutter#69574]
Hi,
I want to create `flutter_test` tests that use a real backend. For this I need to be able to run real HTTP requests from inside the tests.
Right now, all HTTP requests are mocked to return 400 status code.

There was a workaround discussed here: https://github.com/flutter/flutter/issues/35318
The workaround was about adding `HttpOverrides.global = null;`, but it seems this no longer works.

Also, I think this may be possible with `flutter_driver`, but it seems that `flutter_driver` will be deprecated according to this comment: https://github.com/flutter/flutter/issues/7474#issuecomment-558882182. Can you also confirm this?

Any plans to support real HTTP calls in `flutter_test`?

Example code:
```
import 'dart:io';

import 'package:http/http.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  setUpAll(() {
    HttpOverrides.global = null;
  });

  test("Simple test - HTTP request", () async {
    final result = await get(Uri.http("flutter.dev", "/"));
    expect(result.statusCode, 200);
  });

  testWidgets("Widget test - HTTP request", (tester) async {
    final result = await get(Uri.http("flutter.dev", "/"));
    expect(result.statusCode, 200);
  });
}
```

Add this code to a `test.dart` file and try to run it like this: `flutter test -t test/test.dart`. `Widget test - HTTP request` test never finished, it goes in an infinite loop.

`flutter -v doctor`
```
[✓] Flutter (Channel stable, 1.22.3, on Microsoft Windows [Version 10.0.18363.1082], locale en-US)
    • Flutter version 1.22.3 at C:\dev\flutter
    • Framework revision 8874f21e79 (4 days ago), 2020-10-29 14:14:35 -0700
    • Engine revision a1440ca392
    • Dart version 2.10.3


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at C:\dev\android
    • Platform android-29, build-tools 29.0.3
    • ANDROID_SDK_ROOT = C:\dev\android
    • Java binary at: C:\Users\Tudor (Work)\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[!] Android Studio (version 4.0)
    • Android Studio at C:\Users\Tudor (Work)\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[✓] VS Code (version 1.50.1)
    • VS Code at C:\Users\Tudor (Work)\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.16.0

[!] Connected device
    ! Device emulator-5562 is offline.

! Doctor found issues in 2 categories.
```

**Comment by TahaTesser** [comment]
Hi @tudor07 
I can reproduce the issue, the workaround doesn't work on the latest `http
From https://timm.preetz.name/articles/http-request-flutter-test  and https://github.com/flutter/flutter/issues/19588#issuecomment-406771070

<details> 
<summary>flutter doctor -v</summary> 

```bash
import 'dart:io';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
// [...] other testing imports

Future<void> main() async {
  HttpOverrides.global =
      _MyHttpOverrides(); // Setting a customer override that'll use an unmocked HTTP client

  testWidgets(
    'Test with HTTP enabled',
    (tester) async {
      await tester.runAsync(() async {
        // Use `runAsync` to make real asynchronous calls
        expect(
          (await http.Client().get('https://www.google.com/')).statusCode,
          200,
        );
      });
    },
  );
}

class _MyHttpOverrides extends HttpOverrides {}

```

</details>


<details> 
<summary>logs</summary> 

```bash
Running "flutter pub get" in master_flutter...                     377ms
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure object was thrown while running async test code:
  Expected: <200>
  Actual: <400>

When the exception was thrown, this was the stack:
#0      fail (package:test_api/src/frontend/expect.dart:155:31)
#1      _expect (package:test_api/src/frontend/expect.dart:150:3)
#2      expect (package:test_api/src/frontend/expect.dart:59:3)
#3      expect (package:flutter_test/src/widget_tester.dart:429:3)
#4      main.<anonymous closure>.<anonymous closure> (file:///Users/tahatesser/AndroidStudioProjects/master_flutter/test/widget_test.dart:15:9)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)
════════════════════════════════════════════════════════════════════════════════════════════════════
Test failed. See exception logs above.
The test description was: Test with HTTP enabled

Warning: At least one test in this suite creates an HttpClient. When
running a test suite that uses TestWidgetsFlutterBinding, all HTTP
requests will return status code 400, and no network request will
actually be made. Any test expecting a real network connection and
status code will fail.
To test code that needs an HttpClient, provide your own HttpClient
implementation to the code under test, so that your test can
consistently provide a testable response to the code under test.
✖ Test with HTTP enabled
Exited (1)

```

</details>

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 1.22.3, on Mac OS X 10.15.7 19H2, locale en-GB)
    • Flutter version 1.22.3 at /Users/tahatesser/Code/flutter_stable
    • Framework revision 8874f21e79 (4 days ago), 2020-10-29 14:14:35 -0700
    • Engine revision a1440ca392
    • Dart version 2.10.3

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.10.0.rc.1

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (1 available)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E • ios • iOS 14.1

! Doctor found issues in 1 category.
```

```bash
[✓] Flutter (Channel stable, 1.22.3, on Mac OS X 10.15.7 19H2, locale en-GB)
    • Flutter version 1.22.3 at /Users/tahatesser/Code/flutter_stable
    • Framework revision 8874f21e79 (4 days ago), 2020-10-29 14:14:35 -0700
    • Engine revision a1440ca392
    • Dart version 2.10.3

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.10.0.rc.1

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (1 available)
    • Taha’s iPad (mobile) • 00008020-000255113EE8402E • ios • iOS 14.1

! Doctor found issues in 1 category.
```

</details>

**Comment by dnfield** [comment]
I'm not able to reproduce this using only `dart:io`.

I strongly suspect what yo'ure running into is that flutter_test sets up fake async zones, and http requests do real async work. That will get you stuck in your test. However, the following test behaves as expected:

```dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

void main() {
  setUpAll(() {
    HttpOverrides.global = null;
  });
  testWidgets('http', (WidgetTester tester) async {
    await tester.runAsync(() async {
      final HttpClient client = HttpClient();
      final HttpClientRequest request =
          await client.getUrl(Uri.parse('https://google.com'));

      final HttpClientResponse response = await request.close();
      print(response.statusCode);
    });
  });
}
```

Note the `tester.runAsync` block in there.

If you use package:integration_test, you won't have to do the override, but you'll still have to differentiate between real and fake async work.

The problem with not using fake async is your tests become much more prone to timing issues (and, in fact, using network connections at all will make your tests prone to network connectivity and timing issues).

I'm going to close this as WAI, but please feel free to comment if there's something missing and we can reopen.

**Comment by dnfield** [comment]
(without the tester.runAsync block, this test will just hang until it times out)

**Comment by tudor07** [comment]
@dnfield 
Your provided code works, my test for my app does not work.
The only difference that I see is that you use `HttpClient` from `dart:io`, while I use `Client` from `http` library.

My code crashes with no stacktrace at:
`C:\dev\flutter\.pub-cache\hosted\pub.dartlang.org\http-0.12.0+2\lib\src\io_client.dart`
at line 33:
```
var ioRequest = await _inner.openUrl(request.method, request.url);
```

Any reason why `http` library does not work in tests?

**Comment by dnfield** [comment]
I'm not sure - this works for me:

```dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart';

void main() {
  setUpAll(() {
    HttpOverrides.global = null;
  });
  testWidgets('http', (WidgetTester tester) async {
    await tester.runAsync(() async {
      final HttpClient client = HttpClient();
      final HttpClientRequest request =
          await client.getUrl(Uri.parse('https://google.com'));

      final HttpClientResponse response = await request.close();
      print(response.statusCode);
    });
  });

  testWidgets('http2', (WidgetTester tester) async {
    await tester.runAsync(() async {
      final result = await get(Uri.parse('https://google.com'));
      print(result.statusCode);
    });
  });
}
```

There are a large number of things that could cause such failures - in addition to not running in a real async zone, you could be having network connection issues, the server could be down, the process might not have the right permissions on your machine, etc. etc...

In general, for a test, you'll want to mock out HTTP responses. package:http provides some helpers for that in https://pub.dev/documentation/http/latest/testing/testing-library.html.

**Comment by tudor07** [comment]
Mocking is not an option for me.

**Comment by evaisse** [comment]
Hi, 

mocking is not an option for me also. My company has developed a mock proxy that intercept API calls and resolve by using mocks. There is a lot more intelligence than using a basic when() workflow. I would like to use HTTP call only to hit my local http mocking proxy but I can't found any solution to make it work so far. 

I did not have found any exemple to use WidgetTester for making http calls beside the simple exemple above. 
I need to find something that works also by testing the behavior of a widget connected to those async calls.

I did make a working example here by using a basic `Future.value()` solution : 
-  https://github.com/evaisse/demo-flutter-app-widget-tester/blob/master/test/async_widget_test.dart#L75

But using any "real" async call attempts cause the test to fail and the widget not to be renderer as expected...
Example here : 
 - https://github.com/evaisse/demo-flutter-app-widget-tester/blob/master/test/async_widget_test.dart#L94

Same problem if I use a basic `Process.run` : 
 - Future https://github.com/evaisse/demo-flutter-app-widget-tester/blob/master/test/async_widget_test.dart#L113
